File: app/src/main/java/uk/co/sundroid/domain/TimeZoneDetail.java
package uk.co.sundroid.domain;

import android.support.annotation.NonNull;

import java.io.Serializable;
import java.util.TimeZone;

public class TimeZoneDetail implements Comparable<TimeZoneDetail>, Serializable {

	private static final long serialVersionUID = 8968408131775567729L;

	private String id;
	private String cities;
	private TimeZone zone;
	private long currentOffset;
	
	public TimeZoneDetail(String id, String cities, TimeZone zone) {
		this.id = id;
		this.cities = cities;
		this.zone = zone;
		this.currentOffset = zone.getOffset(System.currentTimeMillis());
	}
	
	public String getOffset(long time) {
		
		if (zone.getID().equals("UTC")) {
			return "UTC";
		}
		
		long offset = Math.abs(zone.getOffset(time));
		long offsetHours = offset/(1000 * 60 * 60);
		offset = offset - (offsetHours * (1000 * 60 * 60));
		long offsetMinutes = offset/(1000 * 60);
		
		String offsetHoursStr = "00" + offsetHours;
		offsetHoursStr = offsetHoursStr.substring(offsetHoursStr.length() - 2);
		
		String offsetMinutesStr = "00" + offsetMinutes;
		offsetMinutesStr = offsetMinutesStr.substring(offsetMinutesStr.length() - 2);
		
		String offsetString = offsetHoursStr + ":" + offsetMinutesStr;
		if (zone.getOffset(time) < 0) {
			offsetString = "-" + offsetString;
		} else {
			offsetString = "+" + offsetString;
		}
		if (offsetString.equals("+00:00")) {
			offsetString = "";
		}
		return "GMT" + offsetString;
	}

	public String getId() {
		return id;
	}

	public String getCities() {
		return cities;
	}

	public TimeZone getZone() {
		return zone;
	}

	@Override
	public String toString() {
		return "id=" + id + ", cities=" + cities;
	}

	public int compareTo(@NonNull TimeZoneDetail other) {
		if (this.id.equals("UTC")) {
			return -1;
		} else if (other.id.equals("UTC")) {
			return 1;
		}
		int result = Long.valueOf(this.currentOffset).compareTo(other.currentOffset);
		if (result != 0) {
			return result;
		} else {
			return 1;
		}
	}

}

File: app/src/main/kotlin/uk/co/sundroid/domain/TimeZoneDetail.kt
package uk.co.sundroid.domain

import uk.co.sundroid.util.*
import java.io.Serializable
import java.util.TimeZone

class TimeZoneDetail(val zone: TimeZone, val cities: String?) : Comparable<TimeZoneDetail>, Serializable {

    val id: String
        get() = zone.id

    private val currentOffset: Long = zone.getOffset(System.currentTimeMillis()).toLong()

    fun getOffset(time: Long): String {
        if (zone.id == "UTC") {
            return "UTC"
        }

        val offset = zone.getOffset(time)
        if (offset == 0) {
            return "GMT"
        }

        var absOffset = Math.abs(offset)
        val hours = absOffset / (1000 * 60 * 60)
        absOffset -= hours * (1000 * 60 * 60)
        val minutes = absOffset / (1000 * 60)

        val hoursStr = zeroPad(hours.toString(), 2)
        val minutesStr = zeroPad(minutes.toString(), 2)
        val sign = if (offset < 0) "-" else "+"
        return "GMT$sign$hoursStr:$minutesStr"
    }

    override fun toString(): String {
        return "id=$id, cities=$cities"
    }

    override fun compareTo(other: TimeZoneDetail): Int {
        if (this.id == "UTC") {
            return -1
        } else if (other.id == "UTC") {
            return 1
        }
        return if (this.currentOffset < other.currentOffset) -1 else 1
    }

    companion object {
        private const val serialVersionUID = 8968408131775567729L
    }

}
